nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs

```diff
@@ -0,0 +1,848 @@
+            new_config.max_used_keeper_id,
+            allocator.parent_config.max_used_keeper_id + 1.into()
+        );
+        assert_eq!(new_config.keepers.len(), 5);
+        assert_eq!(
+            new_config.keepers.len(),
+            allocator.parent_config.keepers.len() + 1
+        );
+        assert_eq!(new_config.servers.len(), 2);
+        assert_eq!(
+            new_config.servers.len(),
+            allocator.parent_config.servers.len()
+        );
+
+        // Let's run that plan aginst the new config without changing the
+        // inventory raft config. We should end up with the same config.
+        allocator.parent_config = new_config;
+        allocator.inventory.as_mut().unwrap().leader_committed_log_index += 1;
+        let new_config = allocator.plan().unwrap();
+        assert!(!new_config.needs_generation_bump(&allocator.parent_config));
+
+        // Now let's modify the inventory to reflect that the 5th keeper node
+        // was added to the cluster.
+        //
+        // We should see that the configuration doesn't change because all of
+        // our keeper zones have a keeper that is part of the cluster.
+        allocator.inventory.as_mut().unwrap().raft_config.insert(5.into());
+        allocator.inventory.as_mut().unwrap().leader_committed_log_index += 1;
+        let new_config = allocator.plan().unwrap();
+        assert!(!new_config.needs_generation_bump(&allocator.parent_config));
+    }
+
+    #[test]
+    fn expunge_2_of_5_keeper_zones() {
+        static TEST_NAME: &str = "clickhouse_allocator_expunge_2_of_5_keepers";
+        let logctx = test_setup_log(TEST_NAME);
+
+        let (n_keeper_zones, n_server_zones, n_keepers, n_servers) =
+            (5, 2, 5, 2);
+
+        let (in_service_clickhouse_zones, parent_config) = initial_config(
+            n_keeper_zones,
+            n_server_zones,
+            n_keepers,
+            n_servers,
+        );
+
+        let inventory = Some(ClickhouseKeeperClusterMembership {
+            queried_keeper: KeeperId(1),
+            leader_committed_log_index: 2,
+            raft_config: parent_config.keepers.values().cloned().collect(),
+        });
+
+        let mut allocator = ClickhouseAllocator {
+            log: logctx.log.clone(),
+            in_service_clickhouse_zones,
+            parent_config: parent_config.clone(),
+            inventory,
+        };
+
+        // Our clickhouse cluster config should not have changed
+        // We have 5 keepers and 5 zones and all of them are in the inventory
+        let new_config = allocator.plan().unwrap();
+        assert!(!new_config.needs_generation_bump(&parent_config));
+
+        // Now expunge 2 of the 5 keeper zones by removing them from the
+        // in-service zones
+        allocator.in_service_clickhouse_zones.keepers.pop_first();
+        allocator.in_service_clickhouse_zones.keepers.pop_first();
+        // Bump the inventory index so that we get to this check
```

Comment:
I think this is what I was wondering about above re: `leader_committed_log_index`. It looks like in this case we've only changed the blueprint, but we still need the committed log index to be going up to make the right planning decisions?

Reply:
I changed this behavior in [9401e22](https://github.com/oxidecomputer/omicron/pull/6566/commits/9401e22688052011cb70500791fd0f50fbc60edb) and changed that point in the test to ensure it works as expected. In the other cases I looked at it seems unnecessary to change the tests.